search/retrieval.cpp

```diff
@@ -226,12 +230,46 @@ unique_ptr<coding::CompressedBitVector> RetrieveAddressFeaturesImpl(
   vector<uint64_t> features;
   FeaturesCollector collector(cancellable, features);
 
-  WithSearchTrieRoot<TValue>(context.m_value, [&](TrieRoot<TValue> const & root) {
-    MatchFeaturesInTrie(
-        params, root,
-        [&holder](uint32_t featureIndex) { return !holder.ModifiedOrDeleted(featureIndex); },
-        collector);
-  });
+  // TODO (@y): this code highly depends on the fact that the function
+  // is called on a single-token query params. In any case, this code
+  // must be fixed ASAP, as this is the wrong place for DFA creation.
+  ASSERT_EQUAL(1, params.GetNumTokens(), ());
+
+  for (size_t i = 0; i < params.GetNumTokens(); ++i)
+  {
+    if (params.IsPrefixToken(i))
+    {
+      using DFA = PrefixDFAModifier<UniStringDFA>;
+
+      SearchTrieRequest<DFA> request;
+      for (auto const & sym : params.GetTokens(i))
+        request.m_dfas.emplace_back(UniStringDFA(sym));
+      request.m_langs = params.m_langs;
+
+      WithSearchTrieRoot<TValue>(context.m_value, [&](TrieRoot<TValue> const & root) {
```

Comment:
If you think that this is reasonable, you can move this part before if like that
```
auto const foo =[](TrieRoot<TValue> const & root, Request & request, HolderType & holder, CollectorType & collectot) { ... };
```
And you bind in if and else branch to avoid code duplication.
And please consider writing variables explicitly in capture list.